addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -4173,33 +4173,73 @@ registry.BackgroundShape = SnippetOptionWidget.extend({
     _getShapeData(target = this.$target[0]) {
         const defaultData = {
             shape: '',
-            colors: this._getDefaultColors(),
+            colors: {},
             flip: [],
         };
         const json = target.dataset.oeShapeData;
-        return json ? Object.assign(defaultData, JSON.parse(json.replace(/'/g, '"'))) : defaultData;
+        const shapeData = json ? Object.assign(defaultData, JSON.parse(json.replace(/'/g, '"'))) : defaultData;
+        if (Object.keys(shapeData.colors).length === 0) {
+            shapeData.colors = this._getDefaultColors(shapeData.shape);
+        }
+        return shapeData;
     },
     /**
      * Returns the default colors for the currently selected shape.
      *
      * @private
-     */
-    _getDefaultColors() {
-        const $shapeContainer = this.$target.find('> .o_we_shape')
-            .clone()
-            .addClass('d-none')
-            // Needs to be in document for bg-image class to take effect
-            .appendTo(document.body);
-        const shapeContainer = $shapeContainer[0];
-        $shapeContainer.css('background-image', '');
-        const shapeSrc = shapeContainer && getBgImageURL(shapeContainer);
-        $shapeContainer.remove();
+     * @param {String} shape identifier of the shape to consider instead of the currenlty selected
+     */
+    _getDefaultColors(shape) {
+        let shapeSrc;
+        if (shape) {
+            // locate original in snippet options
+            const $shapeContainer = this.$el.find(".o_we_shape_menu we-button[data-shape='" + shape + "'] div.o_we_shape");
+            if ($shapeContainer.length) {
+                const shapeContainer = $shapeContainer[0];
+                shapeSrc = shapeContainer && getBgImageURL(shapeContainer);
```

Comment:
I don't think this fixes the issue I mentioned:

- user has outdated shape on section
- user changes one of the colors
- this calls `color` which calls `_getShapeData` which calls `_getDefaultColors` with the outdated shape name
- find cannot locate the outdated shape in the shape selector, the method returns an empty object

Basically, in that case, you need to fall back to the original strategy: clone the shape container, remove the color-customized shape with `.css('background-image', '')`, and use the bakground image url that you find then (that's coming from the css rule and has the default colors).

Since you cannot find the shape with default colors in the shape menu reliably, I suggest that you just discard this part of the approach and always go with the clone to avoid needless branching and special cases.

Reply:
The shape-identifier-based request was introduced because it was needed to determine the default colors during shape activation (before the shape actually gets put inside the page content). To avoid the branching, it could be a distinct method instead. The `_getDefaultColors` will then have to be adapted to accept targeting a given section, like `_getShapeData` does, otherwise it returns the default colors of the wrong shape.

Reply:
The branching I'm referring to is that you use two different methods to determine the default colors depending on whether a shape is passed or not, but that is not necessary: you can use the original method in both cases, especially because the new method that you introduced (look for the shape in the shape-selector) is unreliable, so you have to fall back to the original method in those cases either way. At that point, why even look for the shape in the shape selector?

Reply:
Because the implicit palette must be computed before the shape is actually added to the page content.